src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteOutputService.java

```diff
@@ -0,0 +1,623 @@
+    }
+
+    @Override
+    public long getLastChangeTime() throws IOException {
+      return -1;
+    }
+
+    @Override
+    public long getNodeId() throws IOException {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public byte[] getDigest() throws IOException {
+      throw new UnsupportedOperationException();
+    }
+  }
+
+  private static class RegularFileStatus extends DumbFileStatus {
+    private final long size;
+    private final byte[] digest;
+
+    RegularFileStatus(long size, byte[] digest) {
+      this.size = size;
+      this.digest = digest;
+    }
+
+    @Override
+    public boolean isFile() {
+      return true;
+    }
+
+    @Override
+    public boolean isDirectory() {
+      return false;
+    }
+
+    @Override
+    public boolean isSymbolicLink() {
+      return false;
+    }
+
+    @Override
+    public boolean isSpecialFile() {
+      return false;
+    }
+
+    @Override
+    public long getSize() throws IOException {
+      return size;
+    }
+
+    @Override
+    public byte[] getDigest() throws IOException {
+      return digest;
+    }
+  }
+
+  private static class DirectoryFileStatus extends DumbFileStatus {
+    private final long lastModifiedTime;
+
+    DirectoryFileStatus(long lastModifiedTime) {
+      this.lastModifiedTime = lastModifiedTime;
+    }
+
+    @Override
+    public boolean isFile() {
+      return false;
+    }
+
+    @Override
+    public boolean isDirectory() {
+      return true;
+    }
+
+    @Override
+    public boolean isSymbolicLink() {
+      return false;
+    }
+
+    @Override
+    public boolean isSpecialFile() {
+      return false;
+    }
+
+    @Override
+    public long getLastModifiedTime() throws IOException {
+      return this.lastModifiedTime;
+    }
+
+  }
+
+  private static class SymlinkFileStatus extends DumbFileStatus {
+    @Override
+    public boolean isFile() {
+      return false;
+    }
+
+    @Override
+    public boolean isDirectory() {
+      return false;
+    }
+
+    @Override
+    public boolean isSymbolicLink() {
+      return true;
+    }
+
+    @Override
+    public boolean isSpecialFile() {
+      return false;
+    }
+  }
+
+  private class GrpcBatchStat implements BatchStat {
+
+    public List<FileStatusWithDigest> batchStat(boolean includeDigest,
+                                                boolean includeLinks,
+                                                Iterable<PathFragment> paths)
+        throws IOException, InterruptedException {
+      // TODO: Do we need to partition the input, just like in
+      // createSymlinkTree(), or is input already guaranteed to be
+      // bounded in size?
+      BatchStatRequest.Builder builder = BatchStatRequest.newBuilder();
+      builder.setBuildId(stateAcrossBuilds.currentBuildId.toString());
+      builder.setIncludeFileDigest(includeDigest);
+      for (PathFragment path : paths) {
+        builder.addPaths(fixupExecRootPath(path));
```

Comment:
nit: I want to nitpick on that since we make the output service aware of the relativeOutputPath in startBuild, so it technically could know that already. There is an obvious statefulness of the API argument to be made, but I am still curious to your reasoning behind that.

Reply:
At least in the case of bb_clientd, some piece of code needs to convert the paths from absolute to relative to the output path. The reason being that bb_clientd's virtual file system doesn't store files under their absolute names.

I decided to let BatchStat() and BatchCreate() take relative paths, both to make requests smaller, and to reduce ambiguity in the API. These methods can only take paths that reside underneath the output base. For example, you can't call BatchCreate() to construct a file under "/etc" or something.

I think this is also a good thing to do, because the `build.bazel.remote.execution.v2.Output{File,Symlink,Directory}` messages that we embed also contain relative paths.